// argo/argo-jscoverage/frame.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['frame.js']) {
  _$jscoverage['frame.js'] = [];
  _$jscoverage['frame.js'][1] = 0;
  _$jscoverage['frame.js'][2] = 0;
  _$jscoverage['frame.js'][3] = 0;
  _$jscoverage['frame.js'][4] = 0;
  _$jscoverage['frame.js'][5] = 0;
  _$jscoverage['frame.js'][6] = 0;
  _$jscoverage['frame.js'][7] = 0;
}
_$jscoverage['frame.js'][1]++;
module.exports = (function Frame() {
  _$jscoverage['frame.js'][2]++;
  this.url = null;
  _$jscoverage['frame.js'][3]++;
  this.routedResponseHandler = null;
  _$jscoverage['frame.js'][4]++;
  this.routed = false;
  _$jscoverage['frame.js'][5]++;
  this.routeUri = null;
  _$jscoverage['frame.js'][6]++;
  this.oncomplete = null;
  _$jscoverage['frame.js'][7]++;
  this.targetUrl = null;
});
_$jscoverage['frame.js'].source = ["module.exports = function Frame() {","  this.url = null;","  this.routedResponseHandler = null;","  this.routed = false;","  this.routeUri = null;","  this.oncomplete = null;","  this.targetUrl = null;","};"];
